File: backend/app/schemas/projects.py
"""Pydantic schemas for project endpoints, listings, filters, trajectories, and history."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field
from backend.app.schemas.project_month import ProjectMonthObservationRead


class ProjectSummaryItem(BaseModel):
    """Lightweight project summary for list and search results."""

    id: int
    project_code: str
    project_name: str
    agency: str | None = None
    ministry: str | None = None
    sector: str | None = None
    state: str | None = None
    report_month: str

    approval_date: str | None = None
    original_completion_date: str | None = None
    revised_completion_date: str | None = None

    original_cost: float | None = None
    revised_cost: float | None = None
    cumulative_expenditure: float | None = None
    physical_progress: float | None = None

    model_config = ConfigDict(from_attributes=True)


class PaginatedProjectsResponse(BaseModel):
    """Paginated response envelope for project listings."""

    items: list[ProjectSummaryItem]
    total: int
    page: int
    page_size: int
    total_pages: int

    model_config = ConfigDict(from_attributes=True)


class FilterOptionsResponse(BaseModel):
    """Available dynamic filter options derived from current database observations."""

    sectors: list[str] = Field(default_factory=list)
    states: list[str] = Field(default_factory=list)
    agencies: list[str] = Field(default_factory=list)
    ministries: list[str] = Field(default_factory=list)
    report_months: list[str] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)


class QuickSearchResult(BaseModel):
    """Lightweight search result for autocomplete and quick discovery."""

    project_code: str
    project_name: str
    agency: str | None = None
    latest_report_month: str

    model_config = ConfigDict(from_attributes=True)


class ProjectDetailResponse(BaseModel):
    """High-level summary and latest observation for a specific project_code."""

    project_code: str
    project_name: str
    agency: str | None = None
    ministry: str | None = None
    sector: str | None = None
    state: str | None = None
    first_reported_month: str
    latest_report_month: str
    total_observations_count: int
    latest_observation: ProjectMonthObservationRead

    model_config = ConfigDict(from_attributes=True)


class ProjectTrajectoryPoint(BaseModel):
    """Chronological observation point in a project's timeline."""

    report_month: str
    original_cost: float | None = None
    revised_cost: float | None = None
    cumulative_expenditure: float | None = None
    physical_progress: float | None = None

    approval_date: str | None = None
    start_date: str | None = None
    original_completion_date: str | None = None
    revised_completion_date: str | None = None

    agency: str | None = None
    ministry: str | None = None
    sector: str | None = None
    state: str | None = None

    model_config = ConfigDict(from_attributes=True)


class ProjectTrajectoryResponse(BaseModel):
    """Chronological historical trajectory of a project."""

    project_code: str
    project_name: str
    observations_count: int
    trajectory: list[ProjectTrajectoryPoint]

    model_config = ConfigDict(from_attributes=True)


class CostRevisionPoint(BaseModel):
    """Cost observation record for cost revision tracking."""

    report_month: str
    original_cost: float | None = None
    revised_cost: float | None = None
    cumulative_expenditure: float | None = None
    original_cost_raw: str | None = None
    revised_cost_raw: str | None = None
    cumulative_expenditure_raw: str | None = None

    model_config = ConfigDict(from_attributes=True)


class ProjectCostRevisionsResponse(BaseModel):
    """Historical cost revision data for a project."""

    project_code: str
    project_name: str
    revisions: list[CostRevisionPoint]
    latest_original_cost: float | None = None
    latest_revised_cost: float | None = None
    cost_revision_ratio: float | None = None

    model_config = ConfigDict(from_attributes=True)


class ScheduleExtensionPoint(BaseModel):
    """Schedule milestone observation record."""

    report_month: str
    approval_date: str | None = None
    start_date: str | None = None
    original_completion_date: str | None = None
    revised_completion_date: str | None = None
    approval_date_raw: str | None = None
    original_completion_date_raw: str | None = None
    revised_completion_date_raw: str | None = None

    model_config = ConfigDict(from_attributes=True)


class ProjectScheduleExtensionsResponse(BaseModel):
    """Historical schedule timeline and completion date evolution."""

    project_code: str
    project_name: str
    timeline: list[ScheduleExtensionPoint]
    latest_original_completion_date: str | None = None
    latest_revised_completion_date: str | None = None

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/api/projects.ts
/**
 * Project API Endpoints
 */

import { apiClient, API_BASE_URL } from "./client.ts";
import type {
  FilterOptionsResponse,
  PaginatedProjectsResponse,
  ProjectCostRevisionsResponse,
  ProjectDetailResponse,
  ProjectListQueryParams,
  ProjectMonthObservationRead,
  ProjectScheduleExtensionsResponse,
  ProjectTrajectoryResponse,
  QuickSearchResult,
} from "@/types/project.ts";

/**
 * List projects with pagination, sorting, and multi-field filtering.
 */
export async function fetchProjects(params: ProjectListQueryParams = {}): Promise<PaginatedProjectsResponse> {
  return apiClient<PaginatedProjectsResponse>(`${API_BASE_URL}/projects`, {
    params: {
      page: params.page ?? 1,
      page_size: params.page_size ?? 20,
      project_code: params.project_code,
      report_month: params.report_month,
      sector: params.sector,
      state: params.state,
      agency: params.agency,
      ministry: params.ministry,
      search: params.search,
      sort_by: params.sort_by,
      sort_order: params.sort_order,
    },
  });
}

/**
 * Retrieve dynamic filter options (sectors, states, agencies, ministries, months).
 */
export async function fetchFilterOptions(): Promise<FilterOptionsResponse> {
  return apiClient<FilterOptionsResponse>(`${API_BASE_URL}/projects/filters/options`);
}

/**
 * Lightweight autocomplete/quick search for projects.
 */
export async function fetchQuickSearch(query: string, limit = 10): Promise<QuickSearchResult[]> {
  return apiClient<QuickSearchResult[]>(`${API_BASE_URL}/projects/search/quick`, {
    params: { q: query, limit },
  });
}

/**
 * Retrieve project high-level summary and latest observation by project code.
 */
export async function fetchProjectDetail(projectCode: string): Promise<ProjectDetailResponse> {
  return apiClient<ProjectDetailResponse>(`${API_BASE_URL}/projects/${encodeURIComponent(projectCode)}`);
}

/**
 * Retrieve complete 31-field latest observation snapshot.
 */
export async function fetchLatestSnapshot(projectCode: string): Promise<ProjectMonthObservationRead> {
  return apiClient<ProjectMonthObservationRead>(
    `${API_BASE_URL}/projects/${encodeURIComponent(projectCode)}/latest-snapshot`
  );
}

/**
 * Retrieve full chronological timeline of monthly observations.
 */
export async function fetchProjectTrajectory(projectCode: string): Promise<ProjectTrajectoryResponse> {
  return apiClient<ProjectTrajectoryResponse>(
    `${API_BASE_URL}/projects/${encodeURIComponent(projectCode)}/trajectory`
  );
}

/**
 * Retrieve historical original vs revised cost vs cumulative expenditure observations.
 */
export async function fetchCostRevisions(projectCode: string): Promise<ProjectCostRevisionsResponse> {
  return apiClient<ProjectCostRevisionsResponse>(
    `${API_BASE_URL}/projects/${encodeURIComponent(projectCode)}/cost-revisions`
  );
}

/**
 * Retrieve historical milestone schedule extensions and completion dates.
 */
export async function fetchScheduleExtensions(projectCode: string): Promise<ProjectScheduleExtensionsResponse> {
  return apiClient<ProjectScheduleExtensionsResponse>(
    `${API_BASE_URL}/projects/${encodeURIComponent(projectCode)}/schedule-extensions`
  );
}
